core/src/main/java/com/datastax/oss/driver/internal/core/config/cloud/CloudConfigFactory.java

```diff
@@ -225,22 +228,51 @@ protected TrustManagerFactory createTrustManagerFactory(
   @NonNull
   protected BufferedReader fetchProxyMetadata(
       @NonNull URL metadataServiceUrl, @NonNull SSLContext sslContext) throws IOException {
-    try {
-      HttpsURLConnection connection = (HttpsURLConnection) metadataServiceUrl.openConnection();
+    HttpsURLConnection connection = null;
+    int attempt = 0;
+    while (attempt < METADATA_RETRY_MAX_ATTEMPTS) {
+      attempt++;
+      connection = (HttpsURLConnection) metadataServiceUrl.openConnection();
       connection.setSSLSocketFactory(sslContext.getSocketFactory());
       connection.setRequestMethod("GET");
       connection.setRequestProperty("host", "localhost");
-      return new BufferedReader(
-          new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
-    } catch (ConnectException e) {
-      throw new IllegalStateException(
-          "Unable to connect to cloud metadata service. Please make sure your cluster is not parked or terminated",
-          e);
-    } catch (UnknownHostException e) {
-      throw new IllegalStateException(
-          "Unable to resolve host for cloud metadata service. Please make sure your cluster is not terminated",
-          e);
+      // if this is the last attempt, throw
+      // else if the response code is 401, 421, or 5xx, retry
+      // else, throw
+      if (attempt < METADATA_RETRY_MAX_ATTEMPTS
+          && (connection.getResponseCode() == 401
+              || connection.getResponseCode() == 421
+              || connection.getResponseCode() >= 500)) {
```

Comment:
It's not immediately clear to me that this is the extent of response codes we want to handle here.

I agree with retrying on any 5xx error.  It also probably does make sense to retry on a 401 and 421 given what we've seen from this service.  But there are a whole set of [4xx response codes](https://www.rfc-editor.org/rfc/rfc9110#name-client-error-4xx) all of which indicate that the client did something wrong (from the server's perspective).  We could see _any_ of these codes coming off of the metadata service... what do we want to do in that case?  We may need to spend some more time looking at this list and thinking about that.

There are also a lot of [3xx errors](https://www.rfc-editor.org/rfc/rfc9110#name-redirection-3xx) but most of those involve the server telling the client that the thing it's asking for is somewhere else.  In that case retrying that specific request probably doesn't make sense... but I would expect a fully-formed HTTP client to send another (slightly different) request based on the feedback it got from the server.  We're not building a fully-formed HTTP client, though, so I think I'm okay with not retrying 3xx errors for now.

There is part of me wondering if we shouldn't put a real HTTP client in here for this...